src/runtime/cuda/cuda_device_api.cc

```diff
@@ -8,11 +8,38 @@
 #include <dmlc/thread_local.h>
 #include <dgl/runtime/registry.h>
 #include <cuda_runtime.h>
+#include <mutex>
 #include "cuda_common.h"
 
 namespace dgl {
 namespace runtime {
 
+class ThreadSafeWorkspacePool {
+ public:
+  ThreadSafeWorkspacePool(
+      DLDeviceType device_type,
+      std::shared_ptr<DeviceAPI> device) :
+    pool_(device_type, device),
+    mtx_() {
+  }
+
+  void* AllocWorkspace(DGLContext ctx, size_t size) {
+    std::lock_guard<std::mutex> guard(mtx_);
+    return pool_.AllocWorkspace(ctx, size);
+  }
+
+  void FreeWorkspace(DGLContext ctx, void* ptr) {
+    std::lock_guard<std::mutex> guard(mtx_);
+    pool_.FreeWorkspace(ctx, ptr);
+  }
+
+ private:
+  WorkspacePool pool_;
```

Comment:
is this CUDA workspace pol?

Reply:
Right, this workspace pool only gets created with `kDLGPU` and `CUDADeviceAPI::Global()`.